Lib/asyncio/executor.py

```diff
@@ -0,0 +1,226 @@
+import time
+from inspect import isawaitable
+from typing import (Any, AsyncIterable, Awaitable, Iterable, NamedTuple,
+                    Optional, Protocol, cast)
+
+from .exceptions import CancelledError
+from .futures import Future
+from .locks import Event
+from .queues import Queue, QueueShutDown
+from .tasks import FIRST_COMPLETED, Task, create_task, gather, wait, wait_for
+
+__all__ = (
+    "Executor",
+)
+
+
+class _WorkFunction[R](Protocol):
+    def __call__(self, *args, **kwargs) -> R | Awaitable[R]:
+        ...
+
+
+class _WorkItem[R](NamedTuple):
+    fn: _WorkFunction[R]
+    args: tuple[Any, ...]
+    kwargs: dict[Any, Any]
+    future: Future[R]
+
+
+async def _run_work_item[R](work_item: _WorkItem[R]) -> R:
+    result = work_item.fn(*work_item.args, **work_item.kwargs)
+    if isawaitable(result):
+        result = cast(R, await result)
+    return result
+
+
+async def _worker[R](input_queue: Queue[Optional[_WorkItem[R]]]) -> None:
+    while True:
+        work_item = await input_queue.get()
+        try:
+            if work_item is None:
+                break
+
+            item_future = work_item.future
+            if item_future.cancelled():
+                continue
+
+            try:
+                task = create_task(_run_work_item(work_item))
+                await wait([task, item_future], return_when=FIRST_COMPLETED)
+                if item_future.cancelled():
+                    task.cancel()
+                    continue
+                result = task.result()
+            except BaseException as exception:
+                if not item_future.cancelled():
+                    item_future.set_exception(exception)
+                continue
+            if not item_future.cancelled():
+                item_future.set_result(result)
+        finally:
+            input_queue.task_done()
+
+
+async def _azip(*iterables: Iterable | AsyncIterable) -> AsyncIterable[tuple]:
+    def _as_async_iterable[T](
+        iterable: Iterable[T] | AsyncIterable[T],
+    ) -> AsyncIterable[T]:
+        async def _to_async_iterable(
+            iterable: Iterable[T],
+        ) -> AsyncIterable[T]:
+            for item in iterable:
+                yield item
+
+        if isinstance(iterable, AsyncIterable):
+            return iterable
+        return _to_async_iterable(iterable)
+
+    async_iterables = [_as_async_iterable(iterable) for iterable in iterables]
+    iterators = [aiter(async_iterable) for async_iterable in async_iterables]
+    while True:
+        try:
+            items = await gather(*[anext(iterator) for iterator in iterators])
```

Comment:
```suggestion
            items =[anext(iterator) for iterator in iterators]
```
`await gather()` is not for free; it creates a task per item.
Let's use plain list comprehension here. 
If a user wants concurrency, he could create a task explicitly providing arguments.